app/main.py

```
@@ -1,34 +1,122 @@
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self,
+                 ships: list[tuple[tuple[int, int], tuple[int, int]]]) -> None:
+        self.field = {}
+        self._create_ships(ships)
+        self._validate_field()
+
+    def _validate_field(self) -> None:
+        ships = set(self.field.values())
+
+        if len(ships) != 10:
+            raise ValueError("The field must contain exactly 10 ships.")
+
+        lengths = [len(ship.decks) for ship in ships]
+        expected = sorted([4, 3, 3, 2, 2, 2, 1, 1, 1, 1])
+        if sorted(lengths) != expected:
+            raise ValueError("Incorrect composition of ships.")
+
+        occupied = set(self.field.keys())
+        for (row, column) in occupied:
+            for dr in (-1, 0, 1):
+                for dc in (-1, 0, 1):
+                    if dr == dc == 0:
+                        continue
+                    if (row + dr, column + dc) in occupied:
+                        if (self.field[(row, column)]
+                                != self.field.get((row + dr, column + dc))):
+                            raise ValueError("Ships cannot"
+                                             " touch, even diagonally.")
+
+    def _create_ships(self,
+                      ships: list[tuple[tuple[int, int],
+                                        tuple[int, int]]]
+                      ) -> None:
+
+        for start, end in ships:
+            ship_object = Ship(start, end)
+            for deck in ship_object.decks:
+                self.field[(deck.row, deck.column)] = ship_object
```

When placing ships you blindly write into `self.field[(deck.row, deck.column)] = ship_object` which will silently overwrite an existing ship if two ships overlap. Overlapping ships must be considered invalid; the code should detect an already-occupied coordinate during placement and raise a ValueError (instead of silently overwriting). This makes overlaps explicit and avoids masking errors until later validation.